fish-rust/src/threads.rs

```diff
@@ -130,15 +110,29 @@ fn iothread_perform_cant_wait_ffi(callback: &cxx::SharedPtr<ffi::CppCallback>) {
     });
 }
 
-/// A [`ThreadPool`] or [`Debounce`] work request.
+/// A [`ThreadPool`] work request.
 type WorkItem = Box<dyn FnOnce() + 'static + Send>;
 
-/// The queue of [`WorkItem`]s to be executed on the main thread. This is added to from
-/// [`Debounce::enqueue_main_thread()`] and read from in `iothread_service_main()`.
-///
-/// Since items are enqueued from various background threads then read by the main thread, the work
-/// items must implement `Send`.
-static MAIN_THREAD_QUEUE: Mutex<Vec<WorkItem>> = Mutex::new(Vec::new());
+type DebounceResult = Box<dyn Any + Send>;
+
+#[allow(clippy::type_complexity)]
+struct DebounceCallback(Box<dyn FnOnce(&mut ReaderData, DebounceResult) + 'static>);
+
+// Safety: only used on main thread.
+unsafe impl Send for DebounceCallback {}
+
+struct DebounceItem {
+    callback: DebounceCallback,
+    result: Option<DebounceResult>,
+}
+
+static NEXT_COMPLETION_ID: AtomicU64 = AtomicU64::new(0);
+
+static DEBOUNCE_STARTED: once_cell::sync::Lazy<Mutex<HashMap<u64, DebounceCallback>>> =
```

Comment:
Right; I indadvertently relied on the fact that moving the callback to the other thread and back would work.
Turns out it is guaranteed to work because we synchronize on DEBOUNCE_FINISHED but Rust doesn't know that.